include/grpc++/impl/codegen/thrift_serializer.h

```diff
@@ -0,0 +1,217 @@
+#include <grpc/impl/codegen/slice.h>
+#include <grpc/impl/codegen/slice_buffer.h>
+#include <thrift/protocol/TBinaryProtocol.h>
+#include <thrift/protocol/TCompactProtocol.h>
+#include <thrift/protocol/TProtocolException.h>
+#include <thrift/transport/TBufferTransports.h>
+#include <thrift/transport/TTransportUtils.h>
+
+namespace apache {
+namespace thrift {
+namespace util {
+
+using apache::thrift::protocol::TBinaryProtocolT;
+using apache::thrift::protocol::TCompactProtocolT;
+using apache::thrift::protocol::TMessageType;
+using apache::thrift::protocol::TNetworkBigEndian;
+using apache::thrift::transport::TMemoryBuffer;
+using apache::thrift::transport::TBufferBase;
+using apache::thrift::transport::TTransport;
+
+template <typename Dummy, typename Protocol> class ThriftSerializer {
+public:
+  ThriftSerializer()
+      : prepared_ (false)
+      , last_deserialized_ (false)
+      , serialize_version_ (false) {}
+
+  virtual ~ThriftSerializer() {}
+
+  // Serialize the passed type into the internal buffer
+  // and returns a pointer to internal buffer and its size
+  template <typename T> void Serialize(const T& fields, const uint8_t** serializedBuffer,
+      size_t* serializedLen) {
+     // prepare or reset buffer
+    if (!prepared_ || last_deserialized_) {
+      prepare();
+    } else {
+      buffer_->resetBuffer();
+    }
+    last_deserialized_ = false;
+
+    // if required serialize protocol version
+    if (serialize_version_) {
+      protocol_->writeMessageBegin("", TMessageType(0), 0);
+    }
+
+    // serilaize fields into buffer
+    fields.write(protocol_.get());
+
+    // write the end of message
+    if (serialize_version_) {
+      protocol_->writeMessageEnd();
+    }
+
+    uint8_t* byteBuffer;
+    uint32_t byteBufferSize;
+    buffer_->getBuffer(&byteBuffer, &byteBufferSize);
+    *serializedBuffer = byteBuffer;
+    *serializedLen = byteBufferSize; 
+  }
+
+  // Serialize the passed type into the byte buffer
+  template <typename T> void Serialize(const T& fields, grpc_byte_buffer** bp) {
+
+    const uint8_t* byteBuffer;
+    size_t byteBufferSize;
+
+    Serialize(fields, &byteBuffer, &byteBufferSize);
+
+    gpr_slice slice = gpr_slice_from_copied_buffer((char*)byteBuffer,byteBufferSize);
+
+    *bp = grpc_raw_byte_buffer_create(&slice, 1);
+
+    gpr_slice_unref(slice);
+  }
+
+  // Deserialize the passed char array into  the passed type, returns the number
+  // of bytes that have been consumed from the passed string.
+  template <typename T> uint32_t Deserialize(const uint8_t* serializedBuffer, size_t length,
+      T* fields) {
+    // prepare buffer if necessary
+    if (!prepared_) {
+      prepare();
+    }
+    last_deserialized_ = true;
+
+    //reset buffer transport 
+    buffer_->resetBuffer((uint8_t*)serializedBuffer, length);
+
+    // read the protocol version if necessary
+    if (serialize_version_) {
+      std::string name = "";
+      TMessageType mt = (TMessageType) 0;
```

Comment:
C-style cast? Maybe should be TMessageType(0) if it's supposed to be a constructor, or just mt(0) ?